### PR TITLE
Avoid Circular Deps with MRIReco v0.7

### PR DESCRIPTION
**Pull request by beorostica**

### CODE REVIEW

File: KomaMRIBase/src/datatypes/Sequence.jl [hunk 1]
```diff
@@ -408,15 +408,16 @@ Returns the RF pulses and the delta frequency.
 - `Δf_rf`: (`1-row ::Matrix{Float64}`, `[Hz]`) delta frequency vector
 """
 get_rfs(seq::Sequence, t) = begin
+    ϵ = MIN_RISE_TIME
```
Comment (cncastillo) [inline KomaMRIBase/src/datatypes/Sequence.jl:411]: Incorrect indentation
```suggestion
        ϵ = MIN_RISE_TIME
```

File: KomaMRIBase/src/datatypes/Sequence.jl [hunk 2]
```diff
@@ -408,15 +408,16 @@ Returns the RF pulses and the delta frequency.
...
 	A  = seq.RF.A
 	Δf = seq.RF.Δf
 	#Timings
 	T = seq.RF.T
 	delay = seq.RF.delay
 	T0 = cumsum([0; durs(seq)], dims=1)
-	(sum([⏢(A[1,i], t.-T0[i],sum(T[i]).-2EPS,EPS,EPS,delay[i]) for i=1:length(seq)]),
-	 sum([⏢(Δf[1,i],t.-T0[i],sum(T[i]).-2EPS,EPS,EPS,delay[i]) for i=1:length(seq)])
+	(sum([⏢(A[1,i], t.-T0[i],sum(T[i]).-2ϵ,ϵ,ϵ,delay[i]) for i=1:length(seq)]),
+	 sum([⏢(Δf[1,i],t.-T0[i],sum(T[i]).-2ϵ,ϵ,ϵ,delay[i]) for i=1:length(seq)])
```
Comment (cncastillo) [inline KomaMRIBase/src/datatypes/Sequence.jl:420]: Not following Julia Blue
```suggestion
	(sum([⏢(A[1,i], t .- T0[i], sum(T[i]) .- 2ϵ, ϵ, ϵ, delay[i]) for i=1:length(seq)]),
	 sum([⏢(Δf[1,i], t .- T0[i], sum(T[i]) .- 2ϵ, ϵ, ϵ, delay[i]) for i=1:length(seq)])
```

File: KomaMRIBase/src/timing/TimeStepCalculation.jl [hunk 3]
```diff
@@ -0,0 +1,164 @@
+const MIN_RISE_TIME = 1e-10 #100*eps(1.0) #SmallestFloat
```
Comment (cncastillo) [inline KomaMRIBase/src/timing/TimeStepCalculation.jl:1]:
```suggestion
const MIN_RISE_TIME = 1e-10
```

File: KomaMRICore/src/rawdata/ISMRMRD.jl [hunk 4]
```diff
@@ -139,16 +132,16 @@ function signal_to_raw_data(
...
+        "enc_lim_kspace_encoding_step_2" => KomaMRIBase.Limit(0, 0, 0),                     #min, max, center, e.g. partition encoding number
+        "enc_lim_average"                => KomaMRIBase.Limit(0, 0, 0),                     #min, max, center, e.g. signal average number
+        "enc_lim_slice"                  => KomaMRIBase.Limit(0, 0, 0),                     #min, max, center, e.g. imaging slice number
+        "enc_lim_contrast"               => KomaMRIBase.Limit(0, 0, 0),                     #min, max, center, e.g. echo number in multi-echo
+        "enc_lim_phase"                  => KomaMRIBase.Limit(0, 0, 0),                     #min, max, center, e.g. cardiac phase number
+        "enc_lim_repetition"             => KomaMRIBase.Limit(0, 0, 0),                     #min, max, center, e.g. dynamic number for dynamic scanning
+        "enc_lim_set"                    => KomaMRIBase.Limit(0, 0, 0),                     #min, max, center, e.g. flow encoding set
+        "enc_lim_segment"                => KomaMRIBase.Limit(0, 0, 0),                     #min, max, center, e.g. segment number for segmented acquisition
```
Comment (cncastillo) [inline KomaMRICore/src/rawdata/ISMRMRD.jl:144]: Remove `KomaMRIBase.Limit`.

File: KomaMRICore/src/simulation/SimulatorCore.jl [hunk 5]
```diff
@@ -185,8 +257,9 @@ function simulate(
 )
     #Simulation parameter unpacking, and setting defaults if key is not defined
     sim_params = default_sim_params(sim_params)
+    sampling_params = Dict{String,Any}("Δt"=>sim_params["Δt"], "Δt_rf"=>sim_params["Δt_rf"])
```
Comment (cncastillo) [inline KomaMRICore/src/simulation/SimulatorCore.jl:260]:
```suggestion
```

File: KomaMRICore/src/simulation/SimulatorCore.jl [hunk 6]
```diff
@@ -185,8 +257,9 @@ function simulate(
...
     # Simulation init
-    seqd = discretize(seq; sim_params) # Sampling of Sequence waveforms
+    seqd = discretize(seq; sampling_params) # Sampling of Sequence waveforms
```
Comment (cncastillo) [inline KomaMRICore/src/simulation/SimulatorCore.jl:262]:
```suggestion
    seqd = discretize(seq; sampling_params=sim_params) # Sampling of Sequence waveforms
```

File: KomaMRICore/test/test_files/Utils.jl [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+        ζgxo = ceil(100000*sqrt(Area_gxo/Smax))/100000  # From File: 0.00041999999999999996 (JEMRIS optimal definition for trapezoidal without flat-top)
+        Agxo = Area_gxo/ζgxo                            # From File: 0.015560481134096917
+        Area_ky = Ny*Area_gy
+        Area_gyo = 0.5*Area_ky
+        ζgyo = ceil(100000*sqrt(Area_gyo/Smax))/100000  # From File: 0.00037
+        Agyo = Area_gyo/ζgyo                            # From File: 0.013799413552738015 (JEMRIS optimal definition for trapezoidal without flat-top)
+
+        # For delay (from JEMRIS center-to-center TE definition)
+        ΔD = TE - (Trf/2 + 2*ζgxo + Ny*ζgx + Ny*Tgx/2 + (Ny-1)*ζgy)
+    end
```
Comment (cncastillo) [inline KomaMRICore/test/test_files/Utils.jl:62]:
```suggestion
    # User params
    FOVx = 0.230                    # Set
    Nx = 100                        # Set
    FOVy = 0.230                    # Set
    Ny = 100                        # Set
    TE = 0.1                        # Set
    Δtadc = 0.00001                 # Set
    Smax = 10*1_000_000_000/2π/γ    # Set
    
    # For gradients which moves in the k-space (and ADC)
    Area_adc = Nx/FOVx/γ
    Tgx = Nx*Δtadc                                  # From File: 0.001
    Agx = Area_adc/Tgx                              # From File: 0.010211565230481714
    ζgx = ceil(100000*(Agx/Smax))/100000            # From File: 0.00028 (JEMRIS non-optimal definition for trapezoidal with flat-top)
    Tadc = Tgx - Δtadc                              # From File: 0.00099
    ΔDadc = ζgx + Δtadc/2                           # From File: 0.000285
    Area_gy = 1/FOVy/γ
    ζgy = ceil(100000*sqrt(Area_gy/Smax))/100000    # From File: 5.9999999999999995e-5 (JEMRIS optimal definition for trapezoidal without flat-top)
    Agy = Area_gy/ζgy                               # From File: 0.0017019276167022875
    
    # For first corner in the k-space (half of the covered rectangle in the k-space)
    Area_kx = Area_adc + Agx*ζgx
    Area_gxo = 0.5*Area_kx
    ζgxo = ceil(100000*sqrt(Area_gxo/Smax))/100000  # From File: 0.00041999999999999996 (JEMRIS optimal definition for trapezoidal without flat-top)
    Agxo = Area_gxo/ζgxo                            # From File: 0.015560481134096917
    Area_ky = Ny*Area_gy
    Area_gyo = 0.5*Area_ky
    ζgyo = ceil(100000*sqrt(Area_gyo/Smax))/100000  # From File: 0.00037
    Agyo = Area_gyo/ζgyo                            # From File: 0.013799413552738015 (JEMRIS optimal definition for trapezoidal without flat-top)
    
    # For delay (from JEMRIS center-to-center TE definition)
    ΔD = TE - (Trf/2 + 2*ζgxo + Ny*ζgx + Ny*Tgx/2 + (Ny-1)*ζgy)
```

File: KomaMRIFiles/src/RawAcquisitionData/WriteRawData.jl [hunk 8]
```diff
@@ -1,6 +1,6 @@
 """
-For using the Limit Struct of KomaMRICore when writing with MRIFiles
+For using the Limit Struct of KomaMRIBase when writing with MRIFiles
 """
-insertNode(xs, paramName::String, param::KomaMRICore.Limit) = begin
+insertNode(xs, paramName::String, param::KomaMRIBase.Limit) = begin
     insertNode(xs, paramName, MRIFiles.Limit(param.minimum, param.maximum, param.center))
 end
```
Comment (cncastillo) [inline KomaMRIFiles/src/RawAcquisitionData/WriteRawData.jl:6]: Remove this.

File: KomaMRIPlots/src/ui/DisplayFunctions.jl [hunk 9]
```diff
@@ -154,38 +154,38 @@ julia> plot_seq(seq)
...
-      yaxis="y", 
+      yaxis="y",
       showlegend=true
   )
 	idx = ["Gx" "Gy" "Gz"]
 	N = length(seq)
 	O = size(seq.RF,1)
-	ΔT = KomaMRICore.durs(seq)
+	ΔT = seq.DUR
 	T0 = cumsum([0; ΔT],dims=1)
```
Comment (cncastillo) [inline KomaMRIPlots/src/ui/DisplayFunctions.jl:174]: We do this so many times, that we could add a function `get_block_start`  or something like that.

File: examples/5.fat_sat_low_field/CMRA_signal_dynamics.jl [hunk 10]
```diff
@@ -15,6 +15,7 @@ obj = cardiac_phantom(0)
 sim_method = BlochDict(save_Mz=true)
 sim_params = Dict{String,Any}("return_type"=>"mat", "sim_method"=>sim_method, "gpu"=>false,
                              "Nthreads"=>1, "Δt_rf"=>2e-4)
+sampling_params = Dict{String,Any}("Δt"=>sim_params["Δt"], "Δt_rf"=>sim_params["Δt_rf"])
```
Comment (cncastillo) [inline examples/5.fat_sat_low_field/CMRA_signal_dynamics.jl:18]:
```suggestion
```

File: examples/5.fat_sat_low_field/CMRA_signal_dynamics.jl [hunk 11]
```diff
@@ -31,7 +32,7 @@ for i=eachindex(spin_group)
     add_trace!(p0, p1, row=1, col=1)
     add_trace!(p0, p2, row=2, col=1)
 end
-seqd = KomaMRICore.discretize(seq; sim_params)
+seqd = KomaMRICore.discretize(seq; sampling_params)
```
Comment (cncastillo) [inline examples/5.fat_sat_low_field/CMRA_signal_dynamics.jl:35]:
```suggestion
seqd = KomaMRICore.discretize(seq; sampling_params=sim_params)
```

File: KomaMRICore/src/rawdata/ISMRMRD.jl (hunk 4 above)
Comment (beorostica) [reply KomaMRICore/src/rawdata/ISMRMRD.jl:144]: Addressed in branch `avoid-circular-deps`

File: KomaMRIBase/src/datatypes/Sequence.jl [hunk 12]
```diff
@@ -724,3 +725,12 @@ get_eddy_currents(seq::Sequence; Δt=1, λ=80e-3) = begin
...
 end
+
+"""
+It is the same as defined in MRIFiles without the need to import MRIFiles
+"""
+struct Limit
+    minimum::Int
+    maximum::Int
+    center::Int
+end
```
[the remark this answers is not in the document]
Comment (beorostica) [reply KomaMRIBase/src/datatypes/Sequence.jl:736]: Addressed in branch `avoid-circular-deps`

File: KomaMRIFiles/src/RawAcquisitionData/WriteRawData.jl (hunk 8 above)
Comment (beorostica) [reply KomaMRIFiles/src/RawAcquisitionData/WriteRawData.jl:6]: Addressed in branch `avoid-circular-deps`

File: KomaMRIBase/src/datatypes/Sequence.jl (hunk 1 above)
Comment (beorostica) [reply KomaMRIBase/src/datatypes/Sequence.jl:425]: Addressed in https://github.com/cncastillo/KomaMRI.jl/pull/241/commits/6c2507cf6077ffacc5730b20c24184763332d45f

File: KomaMRIBase/src/datatypes/Sequence.jl (hunk 2 above)
Comment (beorostica) [reply KomaMRIBase/src/datatypes/Sequence.jl]: Addressed in https://github.com/cncastillo/KomaMRI.jl/pull/241/commits/6c2507cf6077ffacc5730b20c24184763332d45f

File: KomaMRIBase/src/timing/TimeStepCalculation.jl (hunk 3 above)
Comment (beorostica) [reply KomaMRIBase/src/timing/TimeStepCalculation.jl]: Addressed in https://github.com/cncastillo/KomaMRI.jl/pull/241/commits/6c2507cf6077ffacc5730b20c24184763332d45f

File: KomaMRICore/src/simulation/SimulatorCore.jl (hunk 5 above)
Comment (beorostica) [reply KomaMRICore/src/simulation/SimulatorCore.jl]: Addressed in https://github.com/cncastillo/KomaMRI.jl/pull/241/commits/72790d5e927932b0a83b571459e14a4602156038

File: KomaMRICore/src/simulation/SimulatorCore.jl (hunk 6 above)
Comment (beorostica) [reply KomaMRICore/src/simulation/SimulatorCore.jl]: Addressed in https://github.com/cncastillo/KomaMRI.jl/commit/72790d5e927932b0a83b571459e14a4602156038

File: examples/5.fat_sat_low_field/CMRA_signal_dynamics.jl (hunk 10 above)
Comment (beorostica) [reply examples/5.fat_sat_low_field/CMRA_signal_dynamics.jl]: Addressed in https://github.com/cncastillo/KomaMRI.jl/commit/72790d5e927932b0a83b571459e14a4602156038

File: examples/5.fat_sat_low_field/CMRA_signal_dynamics.jl (hunk 11 above)
Comment (beorostica) [reply examples/5.fat_sat_low_field/CMRA_signal_dynamics.jl]: Addressed in https://github.com/cncastillo/KomaMRI.jl/commit/72790d5e927932b0a83b571459e14a4602156038

File: KomaMRIPlots/src/ui/DisplayFunctions.jl (hunk 9 above)
Comment (beorostica) [reply KomaMRIPlots/src/ui/DisplayFunctions.jl]: Addressed in https://github.com/cncastillo/KomaMRI.jl/pull/241/commits/5998d0fe80082ea26f62b0efa53cfe4a9fe8c1a0

File: KomaMRICore/test/test_files/Utils.jl (hunk 7 above)
Comment (beorostica) [reply KomaMRICore/test/test_files/Utils.jl]: Addressed in https://github.com/cncastillo/KomaMRI.jl/pull/241/commits/adf3f7259bdf27b7e43c90e5787dc93ea011a018.

Originally, the sequence was generated by JEMRIS:
- The factor of 1_000_000_000 is used to relate the gamma of JEMRIS to that used by Koma.
- The expresion ceil(100000*(x))/100000 is to set the JEMRIS time resolution for gradients. (I changed the ceil() expression now)
- The sqrt is employed to determine the rise and fall time as a function of the area and slew rate of a triangular shape, given by ζ = sqrt(Area/Slew rate) = sqrt((G * ζ)/(G/ζ)).

Additionally, I added more descriptive comments.

**Comment by beorostica** [comment]
Closed because `avoid-circular-deps` is meant to be merged to `master`

**Pull request closed without merging**